core/src/main/java/io/grpc/Metadata.java

```diff
@@ -235,7 +305,25 @@ public boolean containsKey(Key<?> key) {
   public <T> void put(Key<T> key, T value) {
     Preconditions.checkNotNull(key, "key");
     Preconditions.checkNotNull(value, "value");
-    storeAdd(key.name, new MetadataEntry(key, value));
+    if (len() == 0 || len() == cap()) {
+      int desiredCap = len() + (len() >>> 1) + 4;
```

Comment:
@carl-mastrangelo 

I would argue that's more theoretical musings, than relevant to our use case. 1) the memory is not re-used by the jvm 2) we don't expect metadata to grow much. vector, arraylist and so on need to worry about hundreds of thousands, even millions of entries. there the factory makes  a huge difference. we expect 4, 8 or maybe 16 entries...

I would KISS.